packages/story-editor/src/app/rightClickMenu/useLayerSelect.js

```diff
@@ -0,0 +1,111 @@
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import {
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from '@web-stories-wp/react';
+import SAT from 'sat';
+import { __ } from '@web-stories-wp/i18n';
+import { Icons } from '@web-stories-wp/design-system';
+
+/**
+ * Internal dependencies
+ */
+import useStory from '../story/useStory';
+import createPolygon from '../canvas/utils/createPolygon';
+import { useCanvas } from '../canvas';
+import { getDefinitionForType } from '../../elements';
+
+function useLayerSelect({ menuItemProps, menuPosition, isMenuOpen }) {
+  const [isSubmenuOpen, setIsSubmenuOpen] = useState(false);
+  const { nodesById } = useCanvas(({ state: { nodesById } }) => ({
+    nodesById,
+  }));
+  const { currentPage, selectedElements, setSelectedElementsById } = useStory(
+    ({
+      state: { currentPage, selectedElements },
+      actions: { setSelectedElementsById },
+    }) => ({
+      currentPage,
+      selectedElements,
+      setSelectedElementsById,
+    })
+  );
+
+  useEffect(() => {
+    // Close submenu if the menu itself also closes.
+    if (!isMenuOpen) {
+      setIsSubmenuOpen(false);
+    }
+  }, [isMenuOpen]);
+
+  const { x, y } = menuPosition;
+  const getIntersectingElements = useCallback(() => {
+    if (!currentPage?.elements?.length || !Object.keys(nodesById)?.length) {
+      return [];
+    }
+    const clickedPoint = new SAT.Vector(x, y);
+    return currentPage.elements
+      .map((element) => {
+        const { id, rotationAngle } = element;
+        const node = nodesById[id];
+        if (!node) {
+          return null;
+        }
+        const { x: elX, y: elY, width, height } = node.getBoundingClientRect();
+        const elementP = createPolygon(rotationAngle, elX, elY, width, height);
+        return SAT.pointInPolygon(clickedPoint, elementP) ? element : null;
+      })
+      .filter((el) => el);
+  }, [currentPage, x, y, nodesById]);
+
+  const subMenuItems = useMemo(() => {
+    const intersectingElements = getIntersectingElements();
+    return intersectingElements.map((element) => {
+      const { id, isBackground, type } = element;
+      const { LayerContent } = getDefinitionForType(type);
+      return {
+        key: id,
+        supportsIcon: true,
+        icon: selectedElements[0].id === id ? <Icons.CheckmarkSmall /> : null,
+        label: isBackground ? (
+          <span>{__('Background', 'web-stories')}</span>
+        ) : (
+          <LayerContent element={element} />
+        ),
+        onClick: () => setSelectedElementsById({ elementIds: [id] }),
+        ...menuItemProps,
```

Comment:
@BrittanyIRL This PR is still WIP but wanted to tag you here since your pod has mainly worked on the element context menu.
Let me know if you see anything off with adding the sub-menu in this way. I tried to keep similar logic as it currently is: defining the props in the `app` folder and rendering these as needed elsewhere.

Also, the `rightClickMenu` provider file is quite massive currently with 1200+ lines, I'm planning to refactor it into smaller files within this PR, let me know if you have any concerns with that.

Reply:
I'll look when i get back on Tuesday, thank you for the heads up. We would have been happy to help with the menu adds and can take the refactor to make smaller files if you would like.

Reply:
Thanks, Brittany. Sure, I'll add an issue to refactor the provider and assign it to you.